std/src/main/scala/cats/std/map.scala

```diff
@@ -0,0 +1,32 @@
+package cats
+package std
+
+
+trait MapInstances {
+
+  implicit def MapInstance[K]: Traverse[Map[K, ?]] with FlatMap[Map[K, ?]] =
+    new Traverse[Map[K, ?]] with FlatMap[Map[K, ?]] {
+      override def traverse[G[_] : Applicative, A, B](fa: Map[K, A])(f: (A) => G[B]): G[Map[K, B]] = ???
+
+      override def flatMap[A, B](fa: Map[K, A])(f: (A) => Map[K, B]): Map[K, B] =
+        fa.flatMap { case (_, a) => f(a) }
+
+      override def foldLeft[A, B](fa: Map[K, A], b: B)(f: (B, A) => B): B =
+        fa.foldLeft(b) { case (x, (k, a)) => f(x, a)}
+
+      override def foldRight[A, B](fa: Map[K, A], b: B)(f: (A, B) => B): B =
+        fa.foldRight(b) { case ((k, a), z) => f(a, z)}
+
+      override def foldRight[A, B](fa: Map[K, A], b: Lazy[B])(f: (A, Lazy[B]) => B): Lazy[B] = {
+        // we use Lazy.byName(...) to avoid memoizing intermediate values.
+        def loop(as: Map[K, A], b: Lazy[B]): Lazy[B] =
+          as.toList match {
+            case Nil => b
+            case a :: rest => Lazy.byName(f(a._2, foldRight(rest.toMap, b)(f)))
```

Comment:
I call it `rest` too, but your reason is better than mine.
On Wed, Feb 4, 2015 at 09:28 Erik Osheim <notifications@github.com> wrote:

> In std/src/main/scala/cats/std/map.scala
> <https://github.com/non/cats/pull/65#discussion_r24085953>:
>
> > +
> > +      override def flatMap[A, B](fa: Map[K, A])(f: (A) => Map[K, B]): Map[K, B] =
> > +        fa.flatMap { case (_, a) => f(a) }
> > +
> > +      override def foldLeft[A, B](fa: Map[K, A], b: B)(f: (B, A) => B): B =
> > +        fa.foldLeft(b) { case (x, (k, a)) => f(x, a)}
> > +
> > +      override def foldRight[A, B](fa: Map[K, A], b: B)(f: (A, B) => B): B =
> > +        fa.foldRight(b) { case ((k, a), z) => f(a, z)}
> > +
> > +      override def foldRight[A, B](fa: Map[K, A], b: Lazy[B])(f: (A, Lazy[B]) => B): Lazy[B] = {
> > +        // we use Lazy.byName(...) to avoid memoizing intermediate values.
> > +        def loop(as: Map[K, A], b: Lazy[B]): Lazy[B] =
> > +          as.toList match {
> > +            case Nil => b
> > +            case a :: rest => Lazy.byName(f(a._2, foldRight(rest.toMap, b)(f)))
>
> I think that's based on code I wrote.
>
> When I'm deconstructing things I like to avoid the words *head* and *tail*
> since they potentially can be confused with the methods involved. In this
> case, we are splitting a list into a (the head value) and rest (the rest
> of the list).
>
> I guess tastes vary. Interested to hear if this bugs other people too?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/non/cats/pull/65/files#r24085953>.
>